jdaviz/app.py

```diff
@@ -911,36 +913,93 @@ def add_data(self, data, data_label=None, notify_done=True):
+            to generation.
+
         Returns
         -------
-        A string data label used by Glue
-
-        """
-
-        # if path is not a file or already ends in [ ] extension, assume it is a data-label
-        if not os.path.isfile(path) or re.search(r'(.+)(\[(.*?)\])$', path):
-            return path
-        else:
-            assert ext, 'A data extension must be specified'
-            p = pathlib.Path(path)
-            stem = p.stem.split(os.extsep)[0]
-            label = f'{stem}[{ext}]'
-            return label
+        data_label : str
+            A unique data label that at its root is either given by the user at load time
+            or created by jdaviz using a description of the loaded file's type.
+        """
+        data_label = None
+
+        if loaded_object is None:
+            pass
+        elif isinstance(loaded_object, str):
+            # This is either the full file path or the basename or the user input data label
+            if loaded_object.lower().endswith(('.jpg', '.jpeg', '.png')):
+                file_format = loaded_object.lower().split(".")[-1]
+                loaded_object = os.path.splitext(os.path.basename(loaded_object))[0]
+                data_label = f"{loaded_object}[{file_format}]"
+            elif os.path.isfile(loaded_object) or os.path.isdir(loaded_object):
+                # File that is not an image file
+                data_label = os.path.splitext(os.path.basename(loaded_object))[0]
+            else:
+                # Not a file path so assumed to be a data label
+                data_label = loaded_object
+        elif isinstance(loaded_object, fits.hdu.hdulist.HDUList):
+            if hasattr(loaded_object, 'file_name'):
+                data_label = f"{loaded_object.file_name}[HDU object]"
+            else:
+                data_label = "Unknown HDU object"
+        elif isinstance(loaded_object, Spectrum1D):
+            data_label = "Spectrum1D"
+        elif isinstance(loaded_object, NDData):
+            data_label = "NDData"
+        elif isinstance(loaded_object, np.ndarray):
+            data_label = "Array"
+
+        if data_label is None and alt_name is not None and isinstance(alt_name, str):
+            data_label = alt_name
+        elif data_label is None:
+            data_label = "Unknown"
+
+        if check_unique:
+            data_label = self.return_unique_name(data_label, ext=ext)
+
+        return data_label
+
+    def return_unique_name(self, data_label, ext=None):
+        if data_label is None:
+            data_label = "Unknown"
+
+        check_if_dup = re.compile(r"(.*)(\s\(\d*\))")
+        labels = self.data_collection.labels
+        number_of_duplicates = 0
+        for label in labels:
+            if re.fullmatch(check_if_dup, label):
+                label_split = label.split(" ")
+                label_without_dup = " ".join(label_split[:-1])
+                label = label_without_dup
```

Comment:
I'm not sure that this logic makes the same assumptions as the regex or not... this is stripping off everything after the last space in the label, right?

Reply:
This is stripping everything after the last space. So "my data label (42)" would become "my data label", which would then be used to check if the data label being added is the same as "my data label" without the parentheses.

Reply:
ok, so since the regex _requires_ numbers in a parenthesis AND with a space before those AND nothing after, then I think this makes sense.  I originally interpretted `label_without_dup` to mean "label without the part that is duplicated", while it really means "label without the (N) suffix".  Thanks for the clarification!

Reply:
I'm going to add a comment to clarify that code block, I can see how it can be confusing (and will certainly be confusing to me in a few months time).